src/Controls/src/Core/Platform/Android/MultiPageFragmentStateAdapter.cs

```diff
@@ -30,7 +34,132 @@ public override Fragment CreateFragment(int position)
 
 		public override long GetItemId(int position)
 		{
-			return _page.Children[position].GetHashCode();
+			return GetItemIdByPosition(position).ItemId;
+		}
+
+		public override bool ContainsItem(long itemId)
+		{
+			return GetItemByItemId(itemId) != null;
+		}
+
+		ItemKey GetItemIdByPosition(int position)
+		{
+			CheckItemKeys();
+			var page = _page.Children[position];
+			for (var i = 0; i < keys.Count; i++)
+			{
+				ItemKey item = keys[i];
+				if (item.Page == page)
+				{
+					return item;
+				}
+			}
+
+			ItemKey itemKey = new ItemKey(page, (ik) => keys.Remove(ik));
+			keys.Add(itemKey);
+
+			return itemKey;
+		}
+
+		ItemKey? GetItemByItemId(long itemId)
+		{
+			CheckItemKeys();
+			for (var i = 0; i < keys.Count; i++)
+			{
+				ItemKey item = keys[i];
+				if (item.ItemId == itemId)
+				{
+					return item;
+				}
+			}
+
+			return null;
+		}
+
+		void CheckItemKeys()
+		{
+			for (var i = keys.Count - 1; i >= 0; i--)
+			{
+				ItemKey item = keys[i];
+
+				if (!_page.Children.Contains(item.Page))
+				{
+					// Disconnect will remove the ItemKey from the keys list
+					item.Disconnect();
+				}
+			}
+		}
+
+		class ItemKey
+		{
+			Page? _page;
+			Action<ItemKey>? _markInvalid;
+			bool _stableView;
+			object? _platformView;
+
+			public ItemKey(Page page, Action<ItemKey> markInvalid)
+			{
+				_page = page;
+				_markInvalid = markInvalid;
+				_page.HandlerChanging += OnHandlerChanging;
+				_page.HandlerChanged += OnHandlerChanged;
+				ItemId = AView.GenerateViewId();
+			}
+
+			public bool Disconnected => _page != null;
+			public Page? Page => _page;
+			public long ItemId { get; }
+			public void Disconnect()
+			{
+				if (_page == null)
+					return;
+
+				_markInvalid?.Invoke(this);
+
+				if (_page != null)
+				{
+					_page.HandlerChanging -= OnHandlerChanging;
+					_page.HandlerChanged -= OnHandlerChanged;
+					_page = null;
+				}
+
+				_platformView = null;
+			}
+
+			public void SetToStableView()
+			{
+				_stableView = true;
+			}
```

Comment:
Who calls this one?

Reply:
🤦

Added a test and organized the code to be a bit more straight forward when a key becomes set in stone. 

https://github.com/dotnet/maui/compare/7a349efc710d9d336e98cd61e86dc7b5a731815a...cde3e739ce3ab4b7c4b24bbef8feffd1479feb96